pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -391,7 +391,16 @@ class Timestamp(_Timestamp):
             # User passed tzinfo instead of tz; avoid silently ignoring
             tz, tzinfo = tzinfo, None
 
-        if isinstance(ts_input, str):
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        if (isinstance(ts_input, Timestamp) and freq is None and
+                tz is None and unit is None and year is None and
+                month is None and day is None and hour is None and
+                minute is None and second is None and
+                microsecond is None and nanosecond is None and
+                tzinfo is None):
```

Comment:
@jbrockmendel This appears the way to go if we need maximum performance.

Reply:
I'm not sure all of these extra checks are worth adding to improve perf when passing a Timestamp to a Timestamp constructor - @AlexKirko how often are you expecting that to actually happen?

Reply:
>  how often are you expecting that to actually happen?

For internal usage there are a lot of places where we do:

```
if isinstance(obj, (datetime, np.datetime64)):
    obj = Timestamp(obj)
```

That's not exactly the usage being checked here, but could benefit in the same way from an optimized no-other-arguments-passed check.

Reply:
Admittedly, I don't know the library well enough to comment on internal usage.

However, what I've done repeatedly in my own projects when on a deadline is take a Dataframe column or a list and just cast the elements into the type I need, trusting that if it already is that class, the performance loss won't be noticeable in the larger program (I do lots of ad-hoc data science modeling). I don't think it's as much of a problem in production-quality code, but a lot of people I work with use pandas to quickly preprocess data for sklearn.

You tend to rely on being able to cut corners when working with a well-supported package, and, currently, calling Timestamp on a Timestamp is about a 100 000 times slower than the proposed shortcut, which can be a nasty shock for someone expecting to just blitz through type conversions during data wrangling.

We could make the code a bit more compact with what was proposed in `4c9eb70`, which you can look up above, but this incurs about a 25% performance loss. I believe that if we do this, might as well add extra two lines. It's a bit grating but the gain is worth it, I think.

Reply:
can you add a comment to the effect of "we do this verbose thing because cython wont optimize a list comprehension (as of cython 0.29.x)"

Reply:
@jbrockmendel Added the comment you requested.